mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxMutabilityHelper.kt

```diff
@@ -27,31 +25,52 @@ private const val IMMUTABLE_MAP_MESSAGE =
-            prop.isSubtype(ArrayMap::class) -> "You cannot use ArrayMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
-            Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT &&
-                prop.isSubtype(android.util.ArrayMap::class) -> "You cannot use ArrayMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
-            prop.isSubtype(HashMap::class) -> "You cannot use HashMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
-            prop.isSubtype(Function::class, KCallable::class) -> {
-                "You cannot use functions inside MvRx state. Only pure data should be represented: ${prop.name}"
+    fun Field.isSubtype(vararg classes: KClass<*>): Boolean {
+        return classes.any { klass ->
+            return when (val returnType = this.type) {
+                is ParameterizedType -> klass.java.isAssignableFrom(returnType.rawType as Class<*>)
+                is Class -> klass.java.isAssignableFrom(returnType)
+                else -> false
             }
-            else -> null
-        }?.let { throw IllegalArgumentException(it) }
+        }
     }
+
+    java.declaredFields
+        // During tests, jacoco can add a transient field called jacocoData.
+        .filter { !Modifier.isTransient(it.modifiers) }
+        .forEach { prop ->
+            when {
+                !Modifier.isFinal(prop.modifiers) -> "State property ${prop.name} must be a val, not a var."
+                prop.isSubtype(ArrayList::class) -> "You cannot use ArrayList for ${prop.name}.\n$IMMUTABLE_LIST_MESSAGE"
+                prop.isSubtype(SparseArray::class) -> "You cannot use SparseArray for ${prop.name}.\n$IMMUTABLE_LIST_MESSAGE"
+                prop.isSubtype(LongSparseArray::class) -> "You cannot use LongSparseArray for ${prop.name}.\n$IMMUTABLE_LIST_MESSAGE"
+                prop.isSubtype(SparseArrayCompat::class) -> "You cannot use SparseArrayCompat for ${prop.name}.\n$IMMUTABLE_LIST_MESSAGE"
+                prop.isSubtype(ArrayMap::class) -> "You cannot use ArrayMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
+                Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT &&
+                    prop.isSubtype(android.util.ArrayMap::class) -> "You cannot use ArrayMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
+                prop.isSubtype(HashMap::class) -> "You cannot use HashMap for ${prop.name}.\n$IMMUTABLE_MAP_MESSAGE"
+                prop.isSubtype(Function::class, KCallable::class) -> {
+                    "You cannot use functions inside MvRx state. Only pure data should be represented: ${prop.name}"
+                }
+                else -> null
+            }?.let { throw IllegalArgumentException(it) }
+        }
 }
 
+/**
+ * Since we can only use java reflection, this basically duck types a data class.
+ * componentN methods are also used for @PersistState.
+ */
+private val Class<*>.isData: Boolean
+    get() {
+        declaredMethods.firstOrNull { it.name == "component1" } ?: return false
```

Comment:
Very small detail: `find` can be used instead of `firstOrNull`. They do the same. `find` is just read better in this case.

Reply:
@tasomaniac IMO, it's just a matter of whichever convention the codebase uses. I wish there was only 1 in the stdlib since they are the same. In the MvRx codebase, firstOrNull is used so I went for consistency.